doc: added suit topology guide (#18098)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/working_with_nrf/nrf54h/ug_nrf54h20_suit_topology.rst

Update path for elements not dedicated to SUIT recovery:

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/nrf54h20_update_topology_for_non_recovery_elements.png

   SUIT topology for the nRF54H20 SoC update path in the most complex predicted case for elements not dedicated to SUIT recovery.

Update path for elements dedicated to SUIT recovery:

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/nrf54h20_update_topology_for_recovery_elements.png

   SUIT topology for the nRF54H20 SoC update path for elements dedicated to SUIT recovery.

Recovery path
=============

Verification failures (for the SUIT manifest signature, SUIT manifest digest, and images) in the invocation path force the device to reboot into the recovery path.
The recovery path is essentially an invocation path with a specific purpose: to download missing elements (such as images and manifests) and provide them to the SDFW as update candidates.

.. warning::
   To enable support for recovery, you must activate support for the application recovery manifest, see `Manifest provisioning information data structure`_.

>    If the device is in a state requiring recovery and the application recovery manifest is not activated or is damaged, the SDFW will boot from the root manifest.

Reviewer: I must've been not precise enough, but we are getting there :slightly_smiling_face: 
```
If the device is in a state requiring recovery and the application recovery manifest is not activated the SDFW will boot from the root manifest.
If the application recovery manifest is activated, but the application recovery manifest is damaged, the SDFW will NOT boot any local images.
```

'.. _ug_nrf54h20_suit_topology:

> nRF54H20 SUIT Topology

Reviewer 2: ```suggestion
nRF54H20 SUIT topology
```

> nRF54H20 SUIT Topology

Author: 54H suit device configuration (change title)

######################

.. contents::
   :local:
   :depth: 2

The SUIT manifest defines both the installation and invocation logic using SUIT commands, classed either as conditions or directives.
This allows you to shape the logic to meet the end product's requirements, even when using a generic Secure Domain Firmware (SDFW).

In addition to the logic, certain properties of the SUIT subsystem must be exposed as configurable parameters:

* The number of SUIT manifests supported by the end product.
* Update and boot policies for the manifests.
* Attributes of the manifests.

The following sections define these properties for products based on the nRF54H20 SoC, and describe the default configuration.

SUIT manifests supported by the nRF54H20 SoC
********************************************

In the most complex predicted case, there can be 11 manifests in total, assigned across the 3 domains.

> Application Domain

Reviewer 2: ```suggestion
Application domain
```

==================

In the application domain, the following five SUIT manifests are supported:

* The root manifest
* Three local app manifests
* A recovery manifest

Local manifests are manifests that operate on the same set of local resources.
In most cases, using a single manifest is expected to be sufficient.
However, this document describes what is supported for the most complex scenarios, such as an A/B approach where two executable images are managed alongside a separate manifest for non-executable resources (for example, images or other media).

> Radio Domain

Reviewer 2: ```suggestion
Radio domain
```

   SUIT manifests supported by the nRF54H20 SoC in the most complex predicted case

This diagram illustrates the most complex predicted configuration.
You can choose a subset of the manifests supported by the nRF54H20.

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/nrf54h20_suit_manifests_minimal.png

   SUIT manifests enabled in nRF54H20, Application and Radio images, without recovery support

This diagram illustrates the minimal predicted configuration that supports the application and radio domains.

Properties associated to SUIT manifest
**************************************

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/suit_manifest_properties.png

   An example of property values of the SUIT Application Local Manifest

This diagram presents a set of properties associated with the SUIT manifest type, along with *example* values.
These properties are not part of the manifest itself but are configuration parameters stored on the device, defining the boundaries for allowable manifest behavior.
The properties covered in this document include:

* The property that enables the manifest

> * The vendor ID
> * The manifest class ID

Reviewer 2: Should vendor ID and manifest class ID be capitalized to be consistent with the rest of the doc?
```suggestion
* The Vendor ID (VID)
* The Manifest Class ID
```

Author: I don't think so, unless PMT considers them as brands

* The downgrade prevention policy
* The independent updateability policy
* The key ID ranges for SUIT manifest verification and signature verification policy
* The property granting the ability to operate on device resources

The properties of Nordic-controlled manifests are hard-coded into the SDFW and cannot be modified.
However, some properties of user-customizable manifests are configurable using the `Manifest provisioning information data structure`_.

> Vendor ID and Manifest Class ID
> ============================

Reviewer 2: ```suggestion
Vendor ID and Manifest Class ID
===============================
```

Vendor ID and Manifest Class ID allow users to specify the Vendor (for example, ``Nordic``, or ``ACME Corp``) and the Manifest Class.

> In other words, they define the intended usage of the manifest, such as the manifest controlling firmware images associated with the Application Domain.

Reviewer 2: Should "Application Domain" be "application domain"?

Author: non capitalized

Since it is impractical to store and process long, variable-length strings, RFC4122 UUID version 5 (uuid5) is used instead.

There is no strict requirement on how to generate these ID values, as long as they remain unique.

> However, the recommended method for creating a Vendor ID (VID) is ``Vendor ID = UUID5(DNS_PREFIX, vendor domain name)``.

Reviewer 2: ```suggestion
However, the recommended method for creating a VID is ``Vendor ID = UUID5(DNS_PREFIX, vendor domain name)``.
```

The following example shows how to generate a VID for ``nordicsemi.com`` using Python::

   uuid5(uuid.NAMESPACE_DNS, 'nordicsemi.com')

This generates ``7617daa5-71fd-5a85-8f94-e28d735ce9f4``.

> The recommended method for creating a manifest class ID is ``Manifest Class ID = UUID5(Vendor ID, Class-Specific-Information)``.
> The following example shows how to generate a manifest class ID using Python::

Reviewer 2: ```suggestion
The recommended method for creating a Manifest Class ID is ``Manifest Class ID = UUID5(Vendor ID, Class-Specific-Information)``.
The following example shows how to generate a Manifest Class ID using Python::
```

Author: I'll not capitalize manifest class

   uuid5(nordic_vid, 'nRF54H20_sample_app')

Assuming ``nordic_vid`` is equal to ``7617daa5-71fd-5a85-8f94-e28d735ce9f4``, it generates ``08c1b599-55e8-5fbc-9e76-7bc29ce1b04d``.

> How Manifest Class ID - based filtering works

Reviewer 2: ```suggestion
How Manifest Class ID-based filtering works
```

---------------------------------------------

The SDFW has access to a list of Manifest Class IDs supported by the device.
During the update candidate evaluation process, the SDFW compares the Manifest Class ID of each candidate manifest against this list.
If the Manifest Class ID of a candidate manifest is not recognized by the device, the update candidate is rejected.

Expressing Class-specific information
-------------------------------------

> Let's illustrate this with an example.

Reviewer 2: ```suggestion
Let us illustrate this with an example.
```

ACME Corp has two product types in its portfolio: roller shutter motors and light bulbs.
Both product types are powered by the nRF54H20 SoC and share the hardware design of the radio-related part, to the extent that they can share the same set of executable binary images for the radio domain.
At the same time, hardware differences between motors and light bulbs are significant and it makes sense to have a separated set of executable binary images for the application domain.

This means that it is worth considering to assign the same Manifest Class ID for Radio Local Manifest for both products, for example::

   uuid5(acme_vid, 'nRF54H20_radio')

And two distinct Manifest Class ID values for Application Local Manifest, for example::

    uuid5(acme_vid, 'nRF54H20_light_bulb_app')

    uuid5(acme_vid, 'nRF54H20_roller_shutter_app')

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/suit_acme_manifests.png

   Manifest hierarchies for ACME Corp devices.

> This approach allows ACME Corp to share the same Radio Domain update package (executable binaries and SUIT manifest) across both products.

Reviewer 2: _radio domain_??

| Downgrade prevention policy      | downgrade_prevention_disabled        |                                                           |
+----------------------------------+--------------------------------------+-----------------------------------------------------------+
| Independent updateability policy | independent_update_denied            |                                                           |
+----------------------------------+--------------------------------------+-----------------------------------------------------------+
| Signature verification policy    | signature_check_disabled             |                                                           |
+----------------------------------+--------------------------------------+-----------------------------------------------------------+

Properties of user-controlled SUIT manifests fixed in the SDFW implementation
*****************************************************************************

Root manifest
=============

+----------------------------------+--------------------------------------+-----------------------------------------------------------+

> | Property                         | Proposed Value                       | Note                                                      |

Reviewer 2: Should "Value" be changed in lower case at all the instances?
```suggestion
| Property                         | Proposed value                       | Note                                                      |
```

Minimal predicted configuration with application and radio support
==================================================================

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/nrf54h20_suit_manifests_minimal.png

   SUIT manifests enabled in nRF54H20, Application and Radio images, without recovery support

This diagram illustrates the minimal predicted configuration that supports the application and radio domains.

Properties associated to SUIT manifest
**************************************

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/suit_manifest_properties.png

   An example of property values of the SUIT Application Local Manifest

The previous diagram presents a set of properties associated with the SUIT manifest type, along with *example* values.
These properties are not part of the manifest itself but are configuration parameters stored on the device, defining the boundaries for allowable manifest behavior.
The properties covered in this document include:

* The property to enable the manifest
* Vendor ID
* Manifest Class ID

> * Downgrade prevention policy

Author: @tomchy Could you check this?

> * Downgrade prevention policy

Reviewer: This is the same type of mechanism, but the page that you've linked is about MCUboot-only. The `CONFIG_MCUBOOT_DOWNGRADE_PREVENTION` does not make sense in scope of SUIT.
I think we may link to this page to indicate that this is an implementation of the MCUboot feature described there, but I'd prefer to keep all MPI knobs and options in a single document.

* Independent updateability policy
* Key ID ranges for SUIT manifest verification and signature verification policy
* The property granting the ability to operate on device resources

The properties of Nordic-controlled manifests are hard-coded into the SDFW and cannot be modified.
However, selected properties of user-customizable manifests are configurable via the `Manifest provisioning information data structure`_.

Vendor ID and Manifest Class ID
============================

Vendor ID and Manifest Class ID allow users to specify the Vendor (for example, ``Nordic``, or ``ACME Corp``) and the Manifest Class.
In other words, they define the intended usage of the manifest, such as *the manifest controlling firmware images associated with the Application Domain*.

As storing and processing long, variable-length strings is impractical, RFC4122 UUID version 5 (uuid5) is used instead.

There is no strict requirement on how to generate these exact values, as long as they remain unique.
However, the recommended method for creating a Vendor ID is ``Vendor ID = UUID5(DNS_PREFIX, vendor domain name)``, as in the following example::

>    uuid5(uuid.NAMESPACE_DNS, 'nordicsemi.com')

Author: @tomchy Could you check this example?

>    uuid5(uuid.NAMESPACE_DNS, 'nordicsemi.com')

Reviewer: This is a generic formula, not bound to a specific language or value:
```
Vendor ID = UUID5(DNS_PREFIX, vendor domain name)
```
This is an example, showing how to generate VID for `nordicsemi.com` using python:
```
uuid5(uuid.NAMESPACE_DNS, 'nordicsemi.com')
```
The `example` may be more specific/practical IMO. If you find it misleading, you may add some notes about values/syntax used by the example.

This generates ``7617daa5-71fd-5a85-8f94-e28d735ce9f4``.

The recommended method to create a Manifest Class ID is ``Manifest Class ID = UUID5(Vendor ID, Class-Specific-Information)``, as in the following example::

>    uuid5(nordic_vid, 'nRF54H20_sample_app')

Author: @tomchy Could you check this example?

>    uuid5(nordic_vid, 'nRF54H20_sample_app')

Reviewer: Answered in the comment below (line 118).

> Assuming ``nordic_vid`` is equal to ``7617daa5-71fd-5a85-8f94-e28d735ce9f4``, it generates ``08c1b599-55e8-5fbc-9e76-7bc29ce1b04d``.

Author: @tomchy Could you check this?

> Assuming ``nordic_vid`` is equal to ``7617daa5-71fd-5a85-8f94-e28d735ce9f4``, it generates ``08c1b599-55e8-5fbc-9e76-7bc29ce1b04d``.

Reviewer: `Class ID` is generated based on the `Vendor ID` (value generated in the previous section) and the class-specific name/information. You may also say that the `Vendor ID` specifies the so-called namespace for the class ID generation (previously - the namespace for Vendor IDs was global/common for all vendors, the `DNS PREFIX`).
Again, the sample provides hints how to generate the same values, as defined inside the "default" configuration.

How Manifest Class ID - based filtering works
---------------------------------------------

The SDFW has access to a list of Manifest Class ID UUIDs supported by the device.
During the update candidate evaluation process, the SDFW compares the Manifest Class ID of each candidate manifest against this list.
If the Manifest Class ID of a candidate manifest is not recognized by the device, the update candidate is rejected.

Expressing Class-Specific-Information
-------------------------------------

Let's illustrate this with an example.
``ACME Corp`` has two product types in portfolio: for example roller shutter motors and light bulbs.

> Both product types are powered by the nRF54H20 SoC and share the hardware design of the radio-related part, to the extent that they can share the same set of executable binary images for the radio domain.

Author: no, we dont capitalize radio domain anymore

At the same time, hardware differences between motors and light bulbs are significant and it makes sense to have a separated set of executable binary images for the application domain.

It is worth to consider assigning the same Manifest Class ID for Radio Local Manifest for both products, for example::

   uuid5(acme_vid, 'nRF54H20_radio')

> and two distinct Manifest Class ID values for Application Local Manifest, for example::

Reviewer 3: ```suggestion
And two distinct Manifest Class ID values for Application Local Manifest, for example::
```

    uuid5(acme_vid, 'nRF54H20_light_bulb_app')

    uuid5(acme_vid, 'nRF54H20_roller_shutter_app')

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/suit_acme_manifests.png

   Manifest hierarchies for ACME Corp devices.

> This approach allows you to share the same radio domain update package (executable binaries and SUIT manifest) across both products.

Reviewer 3: ```suggestion
This approach allows ACME Corp to share the same Radio Domain update package (executable binaries and SUIT manifest) across both products.
```

Additionally, if an application domain update package intended for the light bulb is accidentally delivered to the roller shutter motor, the motor will reject it due to the unrecognized (unsupported) Manifest Class ID.

Downgrade prevention policy
===========================

Product requirements can impose different downgrade prevention policies for different manifests.

> The following policy values are supported:

Reviewer 3: Should this be "property values"?

* ``downgrade_prevention_disabled``
* ``downgrade_prevention_enabled``

How downgrade prevention works
------------------------------

The ``suit-manifest-sequence-number`` (an element of the SUIT manifest) is a monotonically increasing anti-rollback counter.

> As part of the update candidate evaluation process, the suit-manifest-sequence-number of the candidate manifests is compared with the installed counterpart.

Reviewer 3: ```suggestion
As part of the update candidate evaluation process, the ``suit-manifest-sequence-number`` of the candidate manifest is compared with the installed counterpart.
```

The result of this comparison, along with the associated downgrade prevention policy, determines whether a candidate is accepted or rejected.

> Assuming ``downgrade_prevention_enabled`` is enabled, a candidate manifest will be accepted only if its ``suit-manifest-sequence-number`` is not lower than the ``suit-manifest-sequence-number`` of the installed manifest.

Reviewer 3: "now lower" implies that equal number is accepted. Is that intentional?

Author: @wiba-nordic Are you referring to "not lower"?

Reviewer 3: Yes. I don't know what the intended functionality is, but having "not lower then" is the same as "the same or higher as". So if the candidate manifest is accepted when its `suit-manifest-sequence-number` is the same as the installed manifest's, this is correct.

Independent updateability policy
================================

This policy allows the expression of whether a candidate manifest of a specific Manifest Class ID can be updated independently of its parent.
In some cases, this can be the intended behavior, while in others, the opposite can be desired.

The following policy values are supported:

* ``independent_update_allowed``
* ``independent_update_denied``

How independent updateability policy works
------------------------------------------

The system is designed to allow the expression of an update package as a candidate manifest along with its dependencies (candidate, child manifests).

> Considering the simplest case, an update package composed of just one manifest (for example, a Local App Manifest), the policy ``independent_update_denied`` associated with the Local App Manifest would block the update.

Reviewer 3: ```suggestion
In the simplest case, an update package composed of just one manifest (for example, a Local App Manifest), the policy ``independent_update_denied`` associated with the Local App Manifest would block the update.
```

> This be considered intended behavior because images for the App Core (controlled by the Local App Manifest) can rely on functionality provided by images controlled by other manifests, such as the radio manifest.

Reviewer 3: ```suggestion
This be considered intended behavior because images for the Application Core (controlled by the Local App Manifest) can rely on functionality provided by images controlled by other manifests, such as the radio manifest.
```

Algorithms for verifying user-specific SUIT manifests are beyond the scope of this document.

Signing Authorities
-------------------

Each manifest type in the nRF54H20 SoC is associated with a range of Key IDs that are allowed to be used for signing and verifying the signature of a specific manifest.
The user does not have the ability to override these settings.
The range of Key IDs associated with distinct manifests is hard-coded in the SDFW.

Provisioning of key material
----------------------------

Nordic-related public keys for SUIT manifest verification are embedded in the SDFW.
You can provision the respective public keys as part of the device provisioning process.

The key provisioning process for user-specific keys is beyond the scope of this document.

Signature verification policy
-----------------------------

The policy allows you to verify if a manifest must be verified before processing.
The following signature verification policy values are supported:

* ``signature_check_disabled``
* ``signature_check_enabled_on_update``
* ``signature_check_enabled_on_update_and_boot``

.. note::

>    For SUIT manifests related to the application and radio domains, signature verification must be skipped, regardless of the configured signature verification policy, if the respective domain is in one of the following lifecycle states (LCS):

Reviewer 3: ```suggestion
   For SUIT manifests related to the Application and Radio Domains, signature verification must be skipped, regardless of the configured signature verification policy, if the respective domain is in one of the following lifecycle states (LCS):
```

Reviewer 3: Ah, hadn't picked up on that change.

   * ``LCS_EMPTY``
   * ``LCS_ROT``
   * ``LCS_ROT_DEBUG``

Ability to operate on device resources
==========================================

> As the manifest operates on device resources, such as accessing memory (NVM and/or RAM) and starting the processor, access rights are associated with the manifest's Class ID.

Reviewer 3: ```suggestion
As the manifest operates on device resources, such as accessing memory (whether NVM, RAM, or both) and starting the processor, access rights are associated with the Manifest Class ID.
```

To minimize this risk, recovery-dedicated SUIT manifests cannot be updated together with other manifests.

Update path for elements not dedicated to SUIT recovery:

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/nrf54h20_update_topology_for_non_recovery_elements.png

   SUIT topology for the nRF54H20 update path in the most complex predicted case for elements not dedicated to SUIT recovery.

Update path for elements dedicated to SUIT recovery:

.. figure:: ../../../../../../../modules/doc-internal/sw_arch/suit/images/nrf54h20_update_topology_for_recovery_elements.png

   SUIT topology for the nRF54H20 update path for elements dedicated to SUIT recovery.

Recovery path
=============

Verification failures (for the SUIT manifest signature, SUIT manifest digest, and images) in the invocation path force the device to reboot into the recovery path.
The recovery path is essentially an invocation path with a specific purpose: to download missing elements (such as images and manifests) and provide them to the SDFW as update candidates.

.. warning::

>    To enable support for recovery, support for the application recovery manifest must be activated, see `Manifest provisioning information data structure`_.

Reviewer 3: ```suggestion
   To enable support for recovery, you must activate support for the application recovery manifest, see `Manifest provisioning information data structure`_.
```

+==================================+======================================+===========================================================+
| Independent updateability policy | independent_update_allowed           |                                                           |
+----------------------------------+--------------------------------------+-----------------------------------------------------------+
| Signing Key ID Range             | MANIFEST_PUBKEY_APPLICATION_GEN0-2   | 0x40022100 - 0x40022102                                   |
+----------------------------------+--------------------------------------+-----------------------------------------------------------+
| Resource Access Rights           |                                      | Ability to boot the App Core,                             |
|                                  |                                      | Memory access based on App Domain UICR                    |
+----------------------------------+--------------------------------------+-----------------------------------------------------------+

> Application Local Manifest A, B, Spare

Reviewer: `Local Manifest 1, 2, 3`?

+----------------------------------+--------------------------------------------+-----------------------------------------------------------+
| Downgrade prevention policy      | downgrade_prevention_enabled               |                                                           |
+----------------------------------+--------------------------------------------+-----------------------------------------------------------+
| Independent updateability policy | independent_update_denied                  |                                                           |
+----------------------------------+--------------------------------------------+-----------------------------------------------------------+
| Signature verification policy    | signature_check_enabled_on_update_and_boot |                                                           |
+----------------------------------+--------------------------------------------+-----------------------------------------------------------+
| Resource Access Rights           |                                            | Ability to boot the System Controller,                    |

> |                                  |                                            | Memory access - TBD                                       |

Reviewer: TBD?
Since this is not a manifest that a user can write/field that can configure, maybe it's best to simply skip it and write only about the booting abilities?
